# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
os.getcwd()

'C:\\Users\\lalit\\Desktop\\lalpy\\Big Mountain'

In [4]:
df=pd.read_csv('C:\\Users\\lalit\\Desktop\\lalpy\\Big Mountain\\data\\step3_output.csv')
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         1  
4              104.0            80.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [5]:
df = pd.concat([df, pd.get_dummies(df['state'])], axis=1).drop(['state'], axis =1)
print(df.shape)
df.head()

(325, 60)


Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0          2     2       0       0  ...             0   
1          0          0     0       0       4  ...             0   
2          0          0     0       1       0  ...             0   
3          1          0     2       2       1  ...             0   
4          0          1     2       3       1  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [6]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

In [7]:
X

summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  \
0           3939           2500      1        0.0          0          2     2   
1           2600           1540      0        0.0          0          0     0   
2           2090            294      0        0.0          0          0     0   
3          11500           2300      0        0.0          1          0     2   
4          11100           1800      0        0.0          0          1     2   
..           ...            ...    ...        ...        ...        ...   ...   
320         9500           1000      0        0.0          0          0     0   
321         7428            810      0        0.0          0          0     0   
322         7808           1571      0        0.0          0          0     1   
323         9663            990      0        0.0          0          0     0   
324         9500           1100      0        0.0          0          0     0   

     triple  double  surface  ...  Rhode Island  South Dakota  Tennessee  \
0         0       0        2  ...             0             0          0   
1         0       4        0  ...             0             0          0   
2         1       0        2  ...             0             0          0   
3         2       1        2  ...             0             0          0   
4         3       1        0  ...             0             0          0   
..      ...     ...      ...  ...           ...           ...        ...   
320       1       1        1  ...             0             0          0   
321       1       1        1  ...             0             0          0   
322       1       1        0  ...             0             0          0   
323       1       3        1  ...             0             0          0   
324       2       0        0  ...             0             0          0   

     Utah  Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0       0        0         0           0              0          0        0  
1       0        0         0           0              0          0        0  
2       0        0         0           0              0          0        0  
3       0        0         0           0              0          0        0  
4       0        0         0           0              0          0        0  
..    ...      ...       ...         ...            ...        ...      ...  
320     0        0         0           0              0          0        1  
321     0        0         0           0              0          0        1  
322     0        0         0           0              0          0        1  
323     0        0         0           0              0          0        1  
324     0        0         0           0              0          0        1  

[325 rows x 58 columns]

In [8]:
y

0      85.00000
1      53.00000
2      34.00000
3      89.00000
4      78.00000
         ...   
320    64.16681
321    42.00000
322    59.00000
323    49.00000
324    49.00000
Name: AdultWeekend, Length: 325, dtype: float64

In [15]:
print(X_scaled)

[[-0.16488055  1.41397994  1.51059931 ... -0.11162911 -0.22755205
  -0.15886029]
 [-0.52805807  0.37103452 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 [-0.66638558 -0.98262172 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 ...
 [ 0.88450991  0.40471296 -0.30100503 ... -0.11162911 -0.22755205
   6.29483916]
 [ 1.38764233 -0.22648629 -0.30100503 ... -0.11162911 -0.22755205
   6.29483916]
 [ 1.34343177 -0.10698213 -0.30100503 ... -0.11162911 -0.22755205
   6.29483916]]


## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [12]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [14]:
print(y)

[ 85.          53.          34.          89.          78.
  47.          64.16681004  64.16681004  64.16681004  78.
  75.          64.16681004  64.16681004 159.          59.
  84.          69.          64.16681004  64.16681004  83.
  64.16681004  79.          50.         125.          69.
  85.         179.          64.16681004  64.16681004 158.
  64.16681004  89.          64.16681004  25.          64.16681004
  79.          89.          71.          56.          84.
  64.16681004  65.         139.          64.16681004  64.16681004
  72.          65.          60.          55.          62.
  42.          64.          70.          42.          47.
  32.          47.          64.16681004  81.          62.
  43.          64.16681004  71.          55.          64.16681004
  35.          40.          45.          52.          39.
  40.          46.          30.          43.          55.
  49.          64.16681004  32.          75.          99.
 105.          79.          68.          45.    

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

In [17]:
print(X_train)

[[-0.82641152 -0.32426243 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 [ 0.50261749 -0.25907834 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 [-0.70707014 -0.54154272 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 ...
 [-0.36531982 -0.58499878 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 [-0.37996627 -0.13740137 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 [ 1.43836241  0.49053868 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]]


In [18]:
print(X_test)

[[ 0.75486178  0.41449058 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 [-0.93490369 -0.747959   -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 [-0.29751222 -1.08474346 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 ...
 [-0.83183613 -0.17216622  1.51059931 ... -0.11162911 -0.22755205
  -0.15886029]
 [ 0.93658615 -0.32426243 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]
 [ 1.45056778  0.88164321 -0.30100503 ... -0.11162911 -0.22755205
  -0.15886029]]


In [19]:
print(y_train)

[ 68.          50.          59.          60.          79.
  84.          47.          75.          46.          64.16681004
  49.          64.16681004  87.          58.          58.
  84.         139.          60.          79.          72.
 110.          47.          68.          64.16681004  56.
  52.          49.          64.16681004  64.16681004  45.
  50.          39.          42.          49.          64.16681004
  43.          47.          20.          89.          47.
  59.          64.16681004  38.          59.          64.16681004
  44.          83.          56.          64.16681004  45.
  64.16681004  63.          34.          68.          99.
  92.          60.          64.16681004  52.          64.16681004
  64.16681004  73.          87.          64.16681004  89.
  32.          64.16681004  49.          55.         119.
  65.          25.          79.          64.16681004  55.
  84.          65.          40.          76.          45.
  41.          48.          91.         

In [20]:
print(y_test)

[ 48.          60.          43.          88.          49.
  69.          64.16681004  64.16681004  78.          45.
  99.          81.          64.16681004  55.          59.
  64.16681004  62.          81.          64.16681004  44.
  59.          64.16681004  17.          93.          49.
  71.          47.          64.16681004  92.          64.16681004
  72.          50.          79.          49.          32.
  35.          59.          96.          74.          77.
  64.16681004  64.16681004  64.          49.          43.
 158.         105.          59.          64.          64.16681004
  49.          64.16681004  35.34        90.          59.
  78.          40.          45.          42.          99.
  60.          92.          45.          54.          59.
  62.         115.          64.          78.          85.
  69.          85.          63.2         47.          71.
  74.         135.          75.          40.          79.99
  40.          65.        ]


# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [21]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [22]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [26]:
print(y_pred)

[ 49.99145393  61.98754768  67.91723518  90.37036018  50.41625861
  73.79223518  76.84057502  64.96801643  79.23364143  61.01977424
  83.63403205  74.35375861  65.28832893  57.74828986  60.68286018
  68.15551643  51.80786018  86.39770393  65.15942268  41.86083869
  55.78979377 101.77270393  23.98364143  90.17993049  55.43432502
  76.35082893  40.35864143  69.34301643  83.88598518  65.64086799
  73.26245002  57.67211799  84.64233283  41.67309455  36.88989143
  35.84692268  57.79614143  97.71411018  76.7473133   77.01879768
  68.11059455  66.43286018  69.72582893  58.89379768  47.06176643
 145.31567268 115.7160633   64.21801643  60.12475471  54.5495594
  54.08520393  65.33764533  41.37426643  86.40161018  60.25317268
  77.26098518  36.15600471  41.7395008   36.30688361  94.87817268
  43.29809455  88.61645393  50.66381721  60.75903205  54.31957893
  60.64965705 111.31860236  62.12426643  80.72192268  73.36840705
  68.69457893  74.76440315  60.30590705  63.17114143  66.40551643
  76.887938

## Review Model Outcomes — Iterate over additional models as needed

In [30]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
explained_variance_score(y_test, y_pred)

0.8666984158430627

In [31]:
mean_absolute_error(y_test, y_pred)

5.770459865912899

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [32]:
lm.intercept_

64.44848517730031

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [34]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coeff_df = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff_df.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
total_chairs  2.417708e+13
fastQuads     9.642979e+12
surface       8.728601e+12
New York      7.475794e+12
double        7.464447e+12
triple        7.018160e+12
Michigan      6.944903e+12
Colorado      6.084927e+12
California    5.948040e+12
quad          5.827059e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [35]:
file="data/step3_output.csv"
df=pd.read_csv(file)
X=df.drop(['AdultWeekend','Name','state'], axis=1)
y=df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y=y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
mean_absolute_error(y_test, y_pred)

0.8712706137398633


5.9255703636477515

In [37]:
print(y)

[ 85.          53.          34.          89.          78.
  47.          64.16681004  64.16681004  64.16681004  78.
  75.          64.16681004  64.16681004 159.          59.
  84.          69.          64.16681004  64.16681004  83.
  64.16681004  79.          50.         125.          69.
  85.         179.          64.16681004  64.16681004 158.
  64.16681004  89.          64.16681004  25.          64.16681004
  79.          89.          71.          56.          84.
  64.16681004  65.         139.          64.16681004  64.16681004
  72.          65.          60.          55.          62.
  42.          64.          70.          42.          47.
  32.          47.          64.16681004  81.          62.
  43.          64.16681004  71.          55.          64.16681004
  35.          40.          45.          52.          39.
  40.          46.          30.          43.          55.
  49.          64.16681004  32.          75.          99.
 105.          79.          68.          45.    

In [36]:
coeff_df = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff_df.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
AdultWeekday         20.338173
daysOpenLastYear      1.801708
vertical_drop         1.710902
projectedDaysOpen     1.652010
SkiableTerrain_ac     1.111151
quad                  1.079750
fastQuads             1.072051
NightSkiing_ac        1.012147
summit_elev           1.011071
surface               0.908163

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [39]:
file="data/step3_output.csv"
df=pd.read_csv(file)
X=df.drop(['Name','AdultWeekend','state','summit_elev'], axis=1)
y=df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y=y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
mean_absolute_error(y_test, y_pred)

0.8718157050440074


5.921406319146117

In [42]:
print(y)

[ 85.          53.          34.          89.          78.
  47.          64.16681004  64.16681004  64.16681004  78.
  75.          64.16681004  64.16681004 159.          59.
  84.          69.          64.16681004  64.16681004  83.
  64.16681004  79.          50.         125.          69.
  85.         179.          64.16681004  64.16681004 158.
  64.16681004  89.          64.16681004  25.          64.16681004
  79.          89.          71.          56.          84.
  64.16681004  65.         139.          64.16681004  64.16681004
  72.          65.          60.          55.          62.
  42.          64.          70.          42.          47.
  32.          47.          64.16681004  81.          62.
  43.          64.16681004  71.          55.          64.16681004
  35.          40.          45.          52.          39.
  40.          46.          30.          43.          55.
  49.          64.16681004  32.          75.          99.
 105.          79.          68.          45.    

In [41]:
coeff_df = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])  
coeff_df.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
AdultWeekday         20.278859
daysOpenLastYear      1.662549
projectedDaysOpen     1.566177
vertical_drop         1.400607
clusters              1.265617
SkiableTerrain_ac     1.222313
quad                  1.148899
fastQuads             1.142562
NightSkiing_ac        1.121838
surface               0.935895

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.8666984158430627 |  5.770459865912899|-|
| Model 2. | 0.8712706137398633| 5.9255703636477515|'state'|
| Model 3. | 0.8718157050440074 | 5.921406319146117|'state','summit_elev','base_elev'|

Model 3 will be the best model as the highly correlated columns were eliminated.